[WIP] OCPBUGS-16728: Require Service Deletion for LB Type Updates

gcs278 opened the pull request:
Due to the AWS CCM leaking resources when the load balancer type is changed on a service, the cloud team is now blocking updates to the load balancer type. As a result, the Ingress Operator will require the service to be deleted and recreated when the Ingress Controller's load balancer type changes.

This change introduces a Progressing=True status message when the load balancer type is modified, instructing the user on how to effectuate the change. Additionally, the
`service.beta.kubernetes.io/aws-load-balancer-type` annotation is now unmanaged.

WIP:
- I'd like to confirm interactions with also changing Subnets, or EIPAllocations, or Scope at the same time.

lihongan commented:
@gcs278 tested and found one issue:
After switching we got message 
```
ingresscontroller "nlb" is progressing: IngressControllerProgressing: One or more status conditions indicate progressing: LoadBalancerProgressing=True (OperandsProgressing: One or more managed resources are progressing: The IngressController load balancer type was changed from "NLB" to "Classic".  
To effectuate this change, you must delete the service: `oc -n openshift-ingress delete svc/router-nlb`; the service load-balancer will then be deprovisioned and a new one created. This will most likely cause the new load-balancer to have a different host name and IP address and cause disruption. 
To return to the previous state, you can revert the change to the IngressController: `oc -n openshift-ingress-operator patch ingresscontrollers/nlb --type=merge --patch='{"spec":{"endpointPublishingStrategy":{"type":"LoadBalancerService","loadBalancer":{"providerParameters":{"type":"AWS","aws":{"type":"Classic"}}}}}}'`), DeploymentRollingOut=True (DeploymentRollingOut: Waiting for router deployment rollout to finish: 1 old replica(s) are pending termination......
```
The command to revert the change is wrong, it should be `{"type":"NLB"}` because the LB type was changed from "NLB" to "Classic".

lihongan commented:
Another concern is the ingresscontroller status, should it always show the actual status? If yes, then the `spec` and `status` might be different before deleting the svc to effectuate the change. 
And seeing same thing when changing the scope (Internal<>External)

gcs278 commented:
> @gcs278 tested and found one issue: After switching we got message

@lihongan  Thanks, good catch! Fixed.

> Another concern is the ingresscontroller status, should it always show the actual status? If yes, then the `spec` and `status` might be different before deleting the svc to effectuate the change. And seeing same thing when changing the scope (Internal<>External)

Yea it's a great observation, I just talked about this with @Miciah yesterday. This PR won't address having the LB Type in the status reflect the actual LB Type status. It's complicated, and will require much more refactoring than I'm willing to bite off with this PR. I think we should consider this a technical debt item that we should try to solve later.

The LB Type in the status will mostly be equivalent to the LB Type in the spec (for now).

gcs278 commented:
I've taken the WIP off the PR. I'm pretty content with the approach taken from some ideas of @Miciah (i.e. two disjoint "managed" annotation lists). Some parts of the code actually were simplified and code was removed as a result.

I've also dropped the usage of "auto effectuation" or "manual effectuation" as I don't think this aptly conveys the idea that you need to delete/recreate the service (based on conversations with other team members). I went for `managedAtServiceCreationLBServiceAnnotations` to explicitly describe that the annotations are only managed when the service is created, implying the service will need to be recreated for updates to take affect. I'm open to ideas still though.

gcs278 commented:
e2e-aws-operator unrelated failure:
```
=== RUN   TestAll/serial/TestUpdateDefaultIngressControllerSecret
    util_test.go:540: conflict when updating ingress controller openshift-ingress-operator/default: Operation cannot be fulfilled on ingresscontrollers.operator.openshift.io "default": the object has been modified; please apply your changes to the latest version and try again, retrying...
```
/test e2e-aws-operator

Install failure
/test e2e-aws-operator-techpreview 

unrelated
/test e2e-aws-ovn-upgrade
/test e2e-gcp-ovn

lihongan commented:
@gcs278 Another issue is observed. Because PROXY proxy is enabled on Classic but disabled on NLB, so before deleting the LB service to effectuate the change, I can see the router deployment is rolled out with ENV `ROUTER_USE_PROXY_PROTOCOL` updated and service annotation `service.beta.kubernetes.io/aws-load-balancer-proxy-protocol` is changed, that might be potential problem.

gcs278 commented:
> @gcs278 Another issue is observed. Because PROXY proxy is enabled on Classic but disabled on NLB, so before deleting the LB service to effectuate the change, I can see the router deployment is rolled out with ENV `ROUTER_USE_PROXY_PROTOCOL` updated and service annotation `service.beta.kubernetes.io/aws-load-balancer-proxy-protocol` is changed, that might be potential problem.

Ah good catch. I forgot that somethings rely on the status LB Type to be accurate. I need to think about how to fix this.
/wip

Miciah commented:
/assign

candita commented:
/assign @alebedev87

gcs278 commented:
I feel like I'm getting pretty close, but I did find one last minor bug with changing LB Type which results in a Subnets change that I need to fix. We have mixed status usages in `status.endpointPublishingStrategy`. Newer ones represent the actual status (subnets, scope, eipallocations), while the old ones represent the desired state with defaulting (basically what `spec` should be).

The mix between these two concepts, specifically `setDefaultPublishingStrategy` initializing or not initializing `status.endpointPublishingStrategy.loadBalancer.providerParameters.aws.[classicLoadBalancer|networkLoadBalancer]` based on the DESIRED value, while `subnets` requires these structs based on the ACTUAL value. So there might be a situation where we need both classicLoadBalancer and networkLoadBalancer in the status (unfortunately messy).

I'll get back to this next week.

gcs278 commented:
One more thing to figure out in azure and GCP E2E jobs:
```
{  pods/openshift-ingress-operator_ingress-operator-6576f49ff4-tddkg_ingress-operator.log.gz:2024-10-21T22:54:12.745Z	INFO	operator.init	runtime/panic.go:770	Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference	{"controller": "ingress_controller", "object": {"name":"default","namespace":"openshift-ingress-operator"}, "namespace": "openshift-ingress-operator", "name": "default", "reconcileID": "7336952b-4d3e-41f8-9ef6-67e2d01433a5"}
```

gcs278 commented:
I believe I fixed the panic, it was the new function I introduced `clearIngressControllerInactiveAWSLBTypeParametersStatus` that had little to no nil checking, nor checked for the platform type == AWS.

gcs278 commented:
@alebedev87 @Miciah I'm ready for reviews to start. Sorry this one is a somewhat of a large change, but I did break it up into a couple of commit, some of which are `[WILL SQUASH]` which I will squash later (or you can try to convince me to keep separate, but these don't "work" on their own).

The code is starting to feel a bit messy to me, but I did my best. Having status represent the desired value (old) & actual value (new, subnets & EIPs) has resulted in more code paths to test. Open to ideas, but I did my best, and also tried to explain my thoughts in the commit messages.

melvinjoseph86 commented:
/test e2e-aws-ovn-single-node

melvinjoseph86 commented:
@gcs278 Still when we create the NLB controller and deleting the svc, the controller fall backs to Classic. Delete the service doesnt make the LB type NLB. Only patching the controller with explicit NLB mention, only makes it NLB controller.

melvinjoseph86 commented:
@gcs278 Can you please help rebase this PR, i am not able to generate cluster using the cluster bot (as it is giving error message `pull request openshift/cluster-ingress-operator#1142 needs to be rebased to branch master`), i would like to test for OCPBUGS-38411

gcs278 commented:
/retest

gcs278 commented:
@melvinjoseph86, sorry for the delay.

> @gcs278 Still when we create the NLB controller and deleting the svc, the controller fall backs to Classic. Delete the service doesnt make the LB type NLB. Only patching the controller with explicit NLB mention, only makes it NLB controller.

I'm not quite following. Are you setting IngressConfig `lbType` at all here? Or are you just creating an IngressController with `NLB` specified in the spec, deleting the service, and it falls back to Classic? Could you provide your commands to help me understand the problem?

> @gcs278 Can you please help rebase this PR, i am not able to generate cluster using the cluster bot (as it is giving error message pull request openshift/cluster-ingress-operator#1142 needs to be rebased to branch master), i would like to test for OCPBUGS-38411

Yes done, sorry I didn't see the conflict until now.

melvinjoseph86 commented:
@gcs278 Kindly ignore the above comments. The issue which i try to describe is already mentioned by @lihongan and it was relevant to OCPBUGS-16728 only)
Right now i did a pre merge testing with the PR as a fix for OCPBUGS-38411, and it seems working. But @miheer  already create a PR (https://github.com/openshift/cluster-ingress-operator/pull/1130) for that. So not sure i where i can give a qe-apporved label.

gcs278 commented:
@melvinjoseph86 
> We still see the EIP allocation in classic controller configuration. 

You mean the EIPs in the IngressController's spec?
```
spec:
  clientTLS:
    clientCA:
      name: ""
    clientCertificatePolicy: ""
  endpointPublishingStrategy:
    loadBalancer:
      dnsManagementPolicy: Managed
      providerParameters:
        aws:
          networkLoadBalancer:
            eipAllocations:
            - eipalloc-0f76989c4c5ee935e
            - eipalloc-05c9a60faae7885d5
            - eipalloc-0becad7e761cc5d12
```

That's expected, and since it's under `networkLoadBalancer` and the LB Type is `Classic`, the Ingress Operator ignores it. Looks like the EIP allocations were removed from the service which is good.

Really we should have CEL validation that restricts the API from allowing `spec.endpointPublishingStrategy.loadBalancer.providerParameters.aws.networkLoadBalancer` from being set while LB Type is Classic. In other words, your patch should have been rejected. It's bad UX, but not exactly a bug. It's something we are thinking about fixing later.

But either way, I think either way, it looks like this PR resolves [OCPBUGS-38411](https://issues.redhat.com//browse/OCPBUGS-38411) for 4.18 and your results look fine. Keep in mind that [OCPBUGS-38411](https://issues.redhat.com//browse/OCPBUGS-38411) still applies for 4.17 since we are not planning to backport this PR. So really https://github.com/openshift/cluster-ingress-operator/pull/1130 should target release-4.17 now.

alebedev87 commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -785,29 +790,61 @@ func loadBalancerServiceIsUpgradeable(ic *operatorv1.IngressController, deployme
+
 	if platform.Type == configv1.AWSPlatformType && subnetsAWSEnabled {
 		var (
+			// Tricky: Subnets in the status indicates the *actual* scope.
```
Similar question. Can we read subnets from the service annotation directly?

alebedev87 commented on pkg/operator/controller/ingress/status.go:
```diff
@@ -943,12 +955,38 @@ func computeLoadBalancerProgressingStatus(ic *operatorv1.IngressController, serv
 	}
 }
 
+// clearIngressControllerInactiveAWSLBTypeParametersStatus clears AWS parameters for the inactive load balancer type,
+// unless there is a load balancer type change is pending. When a change is pending, setDefaultPublishingStrategy
+// sets the desired parameters for the new LB type. To avoid removing the desired state, we allow
```
The logic of  `setDefaultPublishingStategy` puzzled me in other reviews too. Do I understand correctly that this function reverts the changes done by `setDefaultPublishingStategy` syncing the status with the actual
 state of the service (good direction)?

alebedev87 replied:
Not for this PR but why can't we just stop updating the status in `setDefaultPublishingStategy` and let `syncIngressControllerStatus` take care of the status exclusively?
 
 If the type of the LB is changed by the user:
- we save the desired state in `spec`
- update the status to instruct the user to recreate the service
- keep the LB type in `status` the same as in the service's annotation

Once the service is manually recreated the reconciliation will be triggered and we will reset the status to the new type taken from the service annotation.

Am I missing some hidden value in what `setDefaultPublishingStategy` is doing w.r.t. to the status?

gcs278 replied:
Right, I originally did this status NLB/CLB parameters clearing in `setDefaultPublishingStategy`, then I noticed during pending Scope changes that `setDefaultPublishingStategy` was only keeping the desired parameters and deleting the actual parameter's status (it doesn't know or care about pending changes, it just know and cares about desired state).

So, we can't just clear the NLB/CLB parameters based on desired state, we need to know the actual state (if a change is pending), so we don't prematurely clear status that represents the actual state.

> Not for this PR but why can't we just stop updating the status in setDefaultPublishingStategy and let syncIngressControllerStatus take care of the status exclusively?

Most of the time yes. The only exception is maybe some defaulting logic, especially in `setDefaultProviderParameters` that may need to be added to or maintained. I wanted to put a comment in `setDefaultPublishingStategy` to prevent future developers from adding logic to it accidentally (like what happened with Floating IP). I might go ahead and do that.

> If the type of the LB is changed by the user:
> we save the desired state in spec
> update the status to instruct the user to recreate the service
> keep the LB type in status the same as in the service's annotation

Yes, in a perfect world, but to be clear, this PR doesn't `keep the LB type in status the same as in the service's annotation`, it actually immediately updates the LB type in the status to the desired state while it's pending. Hongan was asking the same [question](https://github.com/openshift/cluster-ingress-operator/pull/1142#issuecomment-2345315070). API Team said that changing the behavior of a status field is incompatible, and we also use the LB Type status field for defaulting, which cause us to loose our defaulting. So this was a dead end.

> Am I missing some hidden value in what setDefaultPublishingStategy is doing w.r.t. to the status?

Not really, but maybe some in memory defaulting to avoid nil pointer exceptions like I mentioned earlier.

alebedev87 reviewed (commented):
Sorry for the long delay.

gcs278 reviewed (commented):
Thanks for the review. I still need to rebase due to the fact that https://github.com/openshift/cluster-ingress-operator/pull/1147 introduced a `spec` field that effectively a `managedAtServiceCreationLBServiceAnnotations`, but not an annotation. So I will need to refactor.

alebedev87 commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -785,29 +790,61 @@ func loadBalancerServiceIsUpgradeable(ic *operatorv1.IngressController, deployme
+// *desired* state, while some newer fields reflect the *actual* state of the object.
 func loadBalancerServiceIsProgressing(ic *operatorv1.IngressController, service *corev1.Service, platform *configv1.PlatformStatus, subnetsAWSEnabled bool, eipAllocationsAWSEnabled bool) error {
 	var errs []error
+	// Tricky: Scope in the status indicates the *desired* scope.
```
> For scope, the spec won't indicate the full desire (user desire + defaulting), so we can't change this without changing the behavior.

Right, I forgot about the defaulting. Then should the comment say something like:
```suggestion
	// Tricky: Scope in the status indicates the *desired* scope with the defaulting taken into account.
```
?

gcs278 replied:
Yea sure it's tricky language.

I'll clarify either way.

alebedev87 commented:
/lgtm
/approve
/hold

LGTM, holding for @Miciah to have a look.

alebedev87 commented on pkg/operator/controller/ingress/load_balancer_service_test.go:
```diff
@@ -1780,12 +1803,19 @@ func Test_shouldRecreateLoadBalancer(t *testing.T) {
 			}
-			if reason != tc.reason {
-				t.Errorf("expected reason %s, got %s", tc.reason, reason)
+			versionsCmpOpts := []cmp.Option{
```
nit: to be considered only if there will be any other requested changes.
```suggestion
			slicesCmpOpts := []cmp.Option{
```

gcs278 replied:
Good eye, copy pasta, done.

gcs278 commented:
@alebedev87 thanks for the review. Just added a unit test `Test_clearIngressControllerInactiveAWSLBTypeParametersStatus` and fixed your last comment.

alebedev87 commented:
/lgtm

gcs278 commented:
/retest

shellyyang1989 commented:
/retest-required

gcs278 commented:
/retest-required

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -23,8 +24,6 @@ import (
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	kerrors "k8s.io/apimachinery/pkg/util/errors"
-	"k8s.io/apimachinery/pkg/util/sets"
```
This import got erroneously moved.

gcs278 replied:
Fixed.

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -264,6 +265,55 @@ var (
+
+		// Add the annotations that do NOT support mutable scope.
+		for platform, annotation := range InternalLBAnnotations {
+			if _, ok := platformsWithMutableScope[platform]; !ok {
+				for name := range annotation {
+					if result[platform] == nil {
+						result[platform] = sets.New[string]()
+					}
+					result[platform].Insert(name)
+				}
+			}
```
A little easier to read:
```suggestion
			if _, ok := platformsWithMutableScope[platform]; ok {
				continue
			}
			for name := range annotation {
				if result[platform] == nil {
					result[platform] = sets.New[string]()
				}
				result[platform].Insert(name)
			}
```

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -264,6 +265,55 @@ var (
+		}
+
+		// Add the annotations that do NOT support mutable scope.
+		for platform, annotation := range InternalLBAnnotations {
+			if _, ok := platformsWithMutableScope[platform]; !ok {
+				for name := range annotation {
```
```suggestion
		for platform, annotations := range InternalLBAnnotations {
			if _, ok := platformsWithMutableScope[platform]; !ok {
				for name := range annotations {
```

gcs278 replied:
Done.

gcs278 replied:
Done.

Miciah commented on pkg/operator/controller/ingress/load_balancer_service_test.go:
```diff
@@ -1758,12 +1774,19 @@ func Test_shouldRecreateLoadBalancer(t *testing.T) {
+				cmpopts.SortSlices(func(a, b string) bool { return a < b }),
+			}
+			if !cmp.Equal(changedAnnotations, tc.expectChangedAnnotations, sliceCmpOpts...) {
+				t.Errorf("expected reason %s, got %s", tc.expectChangedAnnotations, changedAnnotations)
+			}
+			if !cmp.Equal(changedFields, tc.expectChangedFields, sliceCmpOpts...) {
+				t.Errorf("expected reason %s, got %s", tc.expectChangedFields, changedFields)
```
```suggestion
				t.Errorf("expected changedAnnotations %s, got %s", tc.expectChangedAnnotations, changedAnnotations)
			}
			if !cmp.Equal(changedFields, tc.expectChangedFields, sliceCmpOpts...) {
				t.Errorf("expected changedFields %s, got %s", tc.expectChangedFields, changedFields)
```

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -774,7 +792,7 @@ func loadBalancerServiceIsUpgradeable(ic *operatorv1.IngressController, deployme
 	// any discrepancy (in case modified) between the desired annotation values of the controller
 	// and the current annotation values.
-	changed, updated := loadBalancerServiceAnnotationsChanged(current, desired, managedLoadBalancerServiceAnnotations)
+	changed, updated, _ := loadBalancerServiceAnnotationsChanged(current, desired, managedLBServiceAnnotations)
```
```suggestion
	changed, _, updated := loadBalancerServiceAnnotationsChanged(current, desired, managedLBServiceAnnotations)
```

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -801,29 +804,61 @@ func loadBalancerServiceIsUpgradeable(ic *operatorv1.IngressController, deployme
 	}
 	if wantScope != haveScope {
-		err := fmt.Errorf("The IngressController scope was changed from %q to %q.", haveScope, wantScope)
+		changedMsg := fmt.Sprintf("The IngressController scope was changed from %q to %q.", haveScope, wantScope)
+		err := fmt.Errorf(changedMsg)
```
I don't think we want double format expansion, do we?
```suggestion
		err := fmt.Errorf("The IngressController scope was changed from %q to %q.", haveScope, wantScope)
		changedMsg := err.Error()
```
or
```suggestion
		changedMsg := fmt.Sprintf("The IngressController scope was changed from %q to %q.", haveScope, wantScope)
		err := fmt.Errorf("%s", changedMsg)
```
or `import "errors"` and then
```suggestion
		changedMsg := fmt.Sprintf("The IngressController scope was changed from %q to %q.", haveScope, wantScope)
		err := errors.New(changedMsg)
```

gcs278 replied:
Sorry I forgot to comment - but yes good catch, this was treating `changedMsg` as a format string, which is could be problematic.

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -801,29 +804,61 @@ func loadBalancerServiceIsUpgradeable(ic *operatorv1.IngressController, deployme
 		errs = append(errs, err)
 	}
 
+	haveLBType := getAWSLoadBalancerTypeFromServiceAnnotation(service)
+	if platform.Type == configv1.AWSPlatformType {
+		// Tricky: LB Type in the status indicates the *desired* scope (user desire + defaulting).
+		wantLBType := getAWSLoadBalancerTypeInStatus(ic)
```
This logic is confusing.  Why is `haveLBType` computed outside the `if` statement and `wantLBType` inside?  I think `haveLBType` will give `operatorv1.AWSClassicLoadBalancer` on non-AWS platforms, which is kind of silly.  Moreover, it seems to me that all the AWS-specific code (for LB type, subnets, and EIP) should be inside a single `if platform.Type == configv1.AWSPlatformType {` block.

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -801,29 +804,61 @@ func loadBalancerServiceIsUpgradeable(ic *operatorv1.IngressController, deployme
+		err := fmt.Errorf(changedMsg)
 		if _, ok := platformsWithMutableScope[platform.Type]; !ok {
-			err = fmt.Errorf("%[1]s  To effectuate this change, you must delete the service: `oc -n %[2]s delete svc/%[3]s`; the service load-balancer will then be deprovisioned and a new one created.  This will most likely cause the new load-balancer to have a different host name and IP address from the old one's.  Alternatively, you can revert the change to the IngressController: `oc -n openshift-ingres ...
+			ocPatchRevertCmd := fmt.Sprintf("oc -n openshift-ingress-operator patch ingresscontrollers/%[1]s --type=merge --patch='{\"spec\":{\"endpointPublishingStrategy\":{\"loadBalancer\":{\"scope\":\"%[2]s\"}}}}'", ic.Name, haveScope)
```
```suggestion
			ocPatchRevertCmd := fmt.Sprintf(`oc -n openshift-ingress-operator patch ingresscontrollers/%[1]s --type=merge --patch='{"spec":{"endpointPublishingStrategy":{"loadBalancer":{"scope":"%[2]s"}}}}'`, ic.Name, haveScope)
```

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -801,29 +804,61 @@ func loadBalancerServiceIsUpgradeable(ic *operatorv1.IngressController, deployme
 		if _, ok := platformsWithMutableScope[platform.Type]; !ok {
-			err = fmt.Errorf("%[1]s  To effectuate this change, you must delete the service: `oc -n %[2]s delete svc/%[3]s`; the service load-balancer will then be deprovisioned and a new one created.  This will most likely cause the new load-balancer to have a different host name and IP address from the old one's.  Alternatively, you can revert the change to the IngressController: `oc -n openshift-ingres ...
+			ocPatchRevertCmd := fmt.Sprintf("oc -n openshift-ingress-operator patch ingresscontrollers/%[1]s --type=merge --patch='{\"spec\":{\"endpointPublishingStrategy\":{\"loadBalancer\":{\"scope\":\"%[2]s\"}}}}'", ic.Name, haveScope)
+			err = fmt.Errorf(effectuateMessage(changedMsg, ocPatchRevertCmd, service))
```
I think you have extra format expansion again, here and everywhere where you have `fmt.Errorf(effectuateMessage(...))`.

Miciah commented on pkg/operator/controller/ingress/load_balancer_service_test.go:
```diff
@@ -1200,7 +1200,14 @@ func Test_loadBalancerServiceChanged(t *testing.T) {
+			expect: false,
+		},
+		{
+			description: "if the service.beta.kubernetes.io/aws-load-balancer-type is added",
```
```suggestion
			description: "if the service.beta.kubernetes.io/aws-load-balancer-type annotation is added",
```

gcs278 replied:
Done.

Miciah commented on test/e2e/operator_test.go:
```diff
@@ -1370,19 +1370,23 @@ func TestAWSResourceTagsChanged(t *testing.T) {
 	assertLoadBalancerServiceAnnotationWithPollImmediate(t, kclient, defaultIC, awsLBAdditionalResourceTags, expectedTags)
 }
 
+// TestAWSLBTypeChange verifies that process of changing the IngressController's
```
```suggestion
// TestAWSLBTypeChange verifies the process of changing the IngressController's
```

Miciah commented on pkg/operator/controller/ingress/controller_test.go:
```diff
@@ -1515,6 +1524,20 @@ func Test_IsProxyProtocolNeeded(t *testing.T) {
+		},
+		{
+			description: "loadbalancer strategy with ELB, but a service with NLB shouldn't use PROXY",
+			strategy:    &loadBalancerStrategyWithNLB,
```
```suggestion
			strategy:    &loadBalancerStrategy,
```
based on the description of the test case.

gcs278 replied:
Good catch. Fixed.

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -782,7 +777,11 @@ func IsServiceInternal(service *corev1.Service) bool {
-	want, desired, err := desiredLoadBalancerService(ic, deploymentRef, platform, subnetsAWSEnabled, eipAllocationsAWSEnabled)
+	proxyNeeded, err := IsProxyProtocolNeeded(ic, platform, current)
+	if err != nil {
+		return fmt.Errorf("failed to determine if proxy protocol is needed for ingresscontroller %s/%s: %v", ic.Namespace, ic.Name, err)
```
```suggestion
		return fmt.Errorf("failed to determine if proxy protocol is needed for ingresscontroller %s/%s: %w", ic.Namespace, ic.Name, err)
```

gcs278 replied:
Done.

Miciah commented on pkg/operator/controller/ingress/controller.go:
```diff
@@ -1110,7 +1110,19 @@ func (r *reconciler) ensureIngressController(ci *operatorv1.IngressController, d
+
+	proxyNeeded, err := IsProxyProtocolNeeded(ci, platformStatus, currentLBService)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("failed to determine if proxy protocol is needed for ingresscontroller %s/%s: %v", ci.Namespace, ci.Name, err))
```
```suggestion
		errs = append(errs, fmt.Errorf("failed to determine if proxy protocol is needed for ingresscontroller %s/%s: %w", ci.Namespace, ci.Name, err))
```

gcs278 replied:
Done.

Miciah commented on pkg/operator/controller/ingress/controller_test.go:
```diff
@@ -652,20 +652,6 @@ func TestSetDefaultPublishingStrategyHandlesUpdates(t *testing.T) {
 			expectedIC:              makeIC(spec(elb()), status(elbWithNullParameters())),
 			domainMatchesBaseDomain: true,
 		},
-		{
-			name:                    "loadbalancer type changed from ELB to NLB, with old ELB parameters removal",
-			ic:                      makeIC(spec(nlb()), status(elbWithNullParameters())),
-			expectedResult:          true,
-			expectedIC:              makeIC(spec(nlb()), status(nlbWithNullParameters())),
-			domainMatchesBaseDomain: true,
-		},
-		{
-			name:                    "loadbalancer type changed from NLB to ELB, with old NLB parameters removal",
-			ic:                      makeIC(spec(elb()), status(nlbWithNullParameters())),
-			expectedResult:          true,
-			expectedIC:              makeIC(spec(elb()), status(elbWithNullParameters())),
-			domainMatchesBaseDomain: true,
-		},
```
Is there a reason you removed these test cases rather than changing them to match the new behavior?  Granted, `TestSetDefaultPublishingStrategyHandlesUpdates` is already a bit unwieldy, and you did add plenty of test coverage in this commit, and that should be adequate.

gcs278 commented:
Thanks for the reviews @Miciah, I'm running low on time and don't think I'll be able to get to this today, so I think someone else will have to follow up, or we should merge as is. I have no preference.

Let's be sure to coordinate this with @JoelSpeed and the merging of https://github.com/openshift/cluster-cloud-controller-manager-operator/pull/362 before we merge this PR. This PR only is required when https://github.com/openshift/cluster-cloud-controller-manager-operator/pull/362 is merged.

JoelSpeed commented:
Do we have a plan/timeline for merging this PR, I have an outstanding bug fix waiting on this that I'd like to get closed if possible

gcs278 reviewed (commented):
Thanks for the review @Miciah 

I'm ready for another review when you have a chance. I'm just doing some sanity testing on my end, and reviews of my own, but it shouldn't change too much.

CC: @alebedev87 (worth another look)

gcs278 commented:
> @gcs278 Today i tested this PR and PR362 of Joel using cluster bot for [OCPBUGS-38411](https://issues.redhat.com/browse/OCPBUGS-38411). And my findings are below
> ...
> So even with the fix the EIP details got removed from the ingress controller status, but not from the spec....

@melvinjoseph86  Thanks for testing - yes, that's expected behavior. This reflects a longstanding API design issue rather than a bug introduced in this PR.

The existing API allows both `classicLoadBalancerParameters` and `networkLoadBalancerParameters` to coexist in the spec when switching load balancer types. In hindsight, we should have used CEL validation to require users to clear the inactive parameter struct when changing LB types. However, we can't retroactively add stricter validation now, as that would break backward compatibility for existing workflows.

So while your observation is accurate, I think this is a limitation of our original API design rather than an issue with the current PR.

gcs278 commented:
@candita yes I agree, we need to get this merged as soon as we can.

@Miciah can you approve this? I'll squash once you have approved, then I'll get @alebedev87 to re-lgtm it.

rikatz commented:
/remove-label tide/merge-method-squash

I have realized Grant may want to squash on a different order and reword something. But this may be good for future needs

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -617,47 +670,16 @@ func (r *reconciler) updateLoadBalancerService(current, desired *corev1.Service,
+		changedFields = append(changedFields, "spec.loadBalancerIP")
 	}
-	return false, ""
+	return shouldRecreateLB, changedAnnotations, changedFields
```
One implication of this change is that the operator becomes sensitive to changes in trailing blanks, or between no annotation and annotation with an empty value, which previously got discarded by `getSubnetsFromServiceAnnotation` and `getEIPAllocationsFromServiceAnnotation`, right?  I think that that is all right though.

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -591,8 +644,8 @@ func (r *reconciler) deleteLoadBalancerService(service *corev1.Service, options
-	if shouldRecreateLB, reason := shouldRecreateLoadBalancer(current, desired, platform); shouldRecreateLB && autoDeleteLB {
-		log.Info("deleting and recreating the load balancer because "+reason, "namespace", desired.Namespace, "name", desired.Name)
+	if shouldRecreateLB, changedAnnotations, changedFields := shouldRecreateLoadBalancer(current, desired, platform); shouldRecreateLB && autoDeleteLB {
+		log.Info("deleting and recreating the load balancer", "annotations", changedAnnotations, "fields", changedFields, "namespace", desired.Namespace, "name", desired.Name)
```
Per [Kubernetes logging conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#message-style-guidelines), the first letter should be capitalized, and it would be clearer to indicate in the log message *why* the operator is deleting and recreating the SLB:
```suggestion
		log.Info("Deleting and recreating the load balancer service because annotations or fields changed", "annotations", changedAnnotations, "fields", changedFields, "namespace", desired.Namespace, "name", desired.Name)
```

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -280,6 +278,11 @@ var (
+				// Annotation to set either CLB/ELB or NLB.
+				// This annotation was previously fully managed (https://issues.redhat.com/browse/NE-865),
+				// but now required service recreation due to issues with AWS leaking resources when updating
+				// LB Type without recreating the service (see https://issues.redhat.com/browse/OCPBUGS-16728).
```
```suggestion
				// Annotation to specify either CLB or NLB.
				// This annotation was previously fully managed (https://issues.redhat.com/browse/NE-865),
				// but now requires service recreation due to issues with AWS leaking resources when updating
				// LB Type without recreating the service (see https://issues.redhat.com/browse/OCPBUGS-16728).
```

gcs278 replied:
Done

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -848,12 +851,22 @@ func loadBalancerServiceIsProgressing(ic *operatorv1.IngressController, service
+		haveLBType := getAWSLoadBalancerTypeFromServiceAnnotation(service)
+		wantLBType := getAWSLoadBalancerTypeInStatus(ic)
+
+		if wantLBType != "" && wantLBType != haveLBType {
```
Is it possible for `wantLBType` to be empty?  If it is possible, why do you want to ignore an inequality between `wantLBType` and `haveLBType` in that case?

gcs278 replied:
Good question - no it's not possible for `wantLBType` to be empty. This must have been an left over artifact of my iterations.

`haveLBType` can't be empty either because we [check](https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/ingress/status.go#L1057) in `computeLoadBalancerProgressingStatus` that Service is non-nil.

So - I'll remove.

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -848,12 +851,22 @@ func loadBalancerServiceIsProgressing(ic *operatorv1.IngressController, service
+
+		if wantLBType != "" && wantLBType != haveLBType {
+			patchSpec := fmt.Sprintf(`{"spec":{"endpointPublishingStrategy":{"type":"LoadBalancerService","loadBalancer":{"providerParameters":{"type":"AWS","aws":{"type":"%s"}}}}}}`, haveLBType)
+			err := createEffectuationMessage("load balancer type", string(haveLBType), string(wantLBType), patchSpec, ic, service)
```
Correcting my previous suggestion:
```suggestion
			err := createEffectuationMessage("loadBalancer.providerParameters.aws.type", string(haveLBType), string(wantLBType), patchSpec, ic, service)
```

gcs278 replied:
Right - field name is intended to be some sort of identifier for the field so the message can be descriptive. The field name on it's, `type`, isn't self-descriptive enough. Your suggestion is reasonable. Done.

Miciah commented on pkg/operator/controller/ingress/status.go:
```diff
@@ -1105,7 +1105,7 @@ func updateIngressControllerAWSSubnetStatus(ic *operatorv1.IngressController, se
 func updateIngressControllerAWSEIPAllocationStatus(ic *operatorv1.IngressController, service *corev1.Service) {
 	// Set the eipAllocations status based on the actual service annotation and on the load balancer type `NLB`.
-	switch getAWSLoadBalancerTypeInStatus(ic) {
+	switch getAWSLoadBalancerTypeFromServiceAnnotation(service) {
```
I have a comment about the commit message rather than the code, but GitHub doesn't have an obvious way to comment on the commit message.  

Are these changes related to this part of the commit message?  

> This commit does not alter the function of the LB type in the status,
> which has always (incorrectly) represented the desired state. However,
> with this change, a new scenario is introduced where the desired LB
> type may differ from the actual state. Consequently, the LB type
> status may no longer reliably reflect the actual state as it
> previously did, potentially surprising users who mistakenly
> assumed it always indicated the current state.

If so, the commit message could be clearer.  For example, an additional sentence would help: "For this reason, this commit updates the status update logic for subnets and EIP allocations to use the LB type that is specified in the annotation, rather than the LB type that is reported in status."

gcs278 replied:
Yup - I actually already added something like this locally when I squashed and reworded the commit message to be a little more readable, I added:
```
- Updated various functions to read actual LB type from the service annotation
  instead of status, since status represents the desired LB type (spec + defaults)
  and can now differ from the actual LB type during a pending LB type change.
```
So next push will resolve this.

Miciah commented on pkg/operator/controller/ingress/status_test.go:
```diff
@@ -933,7 +944,7 @@ func Test_computeLoadBalancerProgressingStatus(t *testing.T) {
 				nil,
 			),
-			service:        &corev1.Service{},
+			service:        lbServiceWithNLB,
```
It seems like this and other test cases were technically incorrect, irrespective of the changes in this commit, right?  

Ah, but `updateIngressControllerAWSEIPAllocationStatus` and `updateIngressControllerAWSSubnetStatus` didn't look at the service annotations before this commit, so it didn't really matter if the test inputs had the correct service annotations.  Understanding that, this change to the tests makes sense to do at the same time as you make that change to `updateIngressControllerAWSEIPAllocationStatus` and `updateIngressControllerAWSSubnetStatus`.

gcs278 replied:
Yup exactly, now we need the unit test service input to be configured with the annotation.

gcs278 commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -866,21 +900,20 @@ func loadBalancerServiceIsProgressing(ic *operatorv1.IngressController, service
-			changedMsg := fmt.Sprintf("The IngressController eipAllocations were changed from %q to %q.", haveEIPAllocationsPrettyJson, wantEIPAllocationsPrettyJson)
-			ocPatchRevertCmd := fmt.Sprintf("oc -n openshift-ingress-operator patch ingresscontrollers/%[1]s --type=merge --patch='{\"spec\":{\"endpointPublishingStrategy\":{\"type\":\"LoadBalancerService\",\"loadBalancer\":{\"providerParameters\":{\"type\":\"AWS\",\"aws\":{\"type\":\"%[2]s\",\"%[3]s\":{\"eipAllocations\":%[4]s}}}}}}}'", ic.Name, getAWSLoadBalancerTypeInStatus(ic), "networkLoadBalancer",  ...
-			err := fmt.Errorf("%[1]s  To effectuate this change, you must delete the service: `oc -n %[2]s delete svc/%[3]s`; the service load-balancer will then be deprovisioned and a new one created. This will most likely cause the new load-balancer to have a different host name and IP address and cause disruption. To return to the previous state, you can revert the change to the IngressController: `%[4 ...
+			patchSpec := fmt.Sprintf(`{"spec":{"endpointPublishingStrategy":{"type":"LoadBalancerService","loadBalancer":{"providerParameters":{"type":"AWS","aws":{"type":"%s","networkLoadBalancer":{"eipAllocations":%s}}}}}}}`, getAWSLoadBalancerTypeInStatus(ic), haveEIPAllocationsPatchJson)
```
Good point, fixed.

Miciah commented on pkg/operator/controller/ingress/controller_test.go:
```diff
@@ -1507,11 +1507,20 @@ func Test_IsProxyProtocolNeeded(t *testing.T) {
 				Protocol: operatorv1.ProxyProtocol,
 			},
 		}
+		serviceWithELB = corev1.Service{}
```
As I understand it, CLBs and NLBs are both ELBs.  
```suggestion
		serviceWithCLB = corev1.Service{}
```

gcs278 commented on test/e2e/operator_test.go:
```diff
@@ -1453,50 +1457,81 @@ func TestAWSLBTypeChange(t *testing.T) {
-	if v := lbService.Annotations[ingresscontroller.AWSLBTypeAnnotation]; len(v) != 0 {
-		t.Fatalf("load balancer service has unexpected %s=%s annotation", ingresscontroller.AWSLBTypeAnnotation, v)
-	}
+	// LB Annotation should be empty by default (meaning CLB).
```
Done.

Miciah commented on pkg/operator/controller/ingress/load_balancer_service.go:
```diff
@@ -778,7 +774,11 @@ func IsServiceInternal(service *corev1.Service) bool {
+	if err != nil {
+		return fmt.Errorf("failed to determine if proxy protocol is needed for ingresscontroller %s/%s: %w", ic.Namespace, ic.Name, err)
+	}
+	want, desired, err := desiredLoadBalancerService(ic, deploymentRef, platform, proxyNeeded)
```
This logic is tricky.  Maybe it would make sense to pass a nil service value to `IsProxyProtocolNeeded` so that it flags an issue if the IngressController and the service contradict each other?  But I suppose that that can only happen if the IngressController and service also contradict each other regarding the LB type, right?

Miciah commented on pkg/operator/controller/ingress/status_test.go:
```diff
@@ -2719,6 +2720,44 @@ func Test_IngressStatusesEqual(t *testing.T) {
+					},
+				},
+			},
+		},
```
I'm confused as to what exactly this test case is covering.  Is it the case where the IngressController specifies an NLB but also specifies parameters for an CLB, and those CLB parameters get cleared?  Something like the following?
```suggestion
		{
			description: "networkLoadBalancer parameters cleared",
			expected:    false,
			a: operatorv1.IngressControllerStatus{
				EndpointPublishingStrategy: &operatorv1.EndpointPublishingStrategy{
					Type: operatorv1.LoadBalancerServiceStrategyType,
					LoadBalancer: &operatorv1.LoadBalancerStrategy{
						ProviderParameters: &operatorv1.ProviderLoadBalancerParameters{
							AWS: &operatorv1.AWSLoadBalancerParameters{
								Type:                          operatorv1.AWSNetworkLoadBalancer,
								ClassicLoadBalancerParameters: &operatorv1.AWSClassicLoadBalancerParameters{
									Subnets: &operatorv1.AWSSubnets{
										Names: []operatorv1.AWSSubnetName{"name-890123", "name-123456"},
									},
								},
								NetworkLoadBalancerParameters: &operatorv1.AWSNetworkLoadBalancerParameters{},
							},
						},
					},
				},
			},
			b: operatorv1.IngressControllerStatus{
				EndpointPublishingStrategy: &operatorv1.EndpointPublishingStrategy{
					Type: operatorv1.LoadBalancerServiceStrategyType,
					LoadBalancer: &operatorv1.LoadBalancerStrategy{
						ProviderParameters: &operatorv1.ProviderLoadBalancerParameters{
							AWS: &operatorv1.AWSLoadBalancerParameters{
								Type:                          operatorv1.AWSNetworkLoadBalancer,
								NetworkLoadBalancerParameters: &operatorv1.AWSNetworkLoadBalancerParameters{},
							},
						},
					},
				},
			},
		},
```
